README.md
# js-cid  <!-- omit in toc -->

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://protocol.ai/)
[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](http://webchat.freenode.net/?channels=%23ipfs)
[![Travis CI](https://flat.badgen.net/travis/multiformats/js-cid)](https://travis-ci.com/multiformats/js-cid)
[![Coverage Status](https://coveralls.io/repos/github/multiformats/js-cid/badge.svg?branch=master)](https://coveralls.io/github/multiformats/js-cid?branch=master)
[![Dependency Status](https://david-dm.org/multiformats/js-cid.svg?style=flat-square)](https://david-dm.org/multiformats/js-cid)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
[![Greenkeeper badge](https://badges.greenkeeper.io/multiformats/js-cid.svg)](https://greenkeeper.io/)

> [CID](https://github.com/multiformats/cid) implementation in JavaScript.

## Lead Maintainer   <!-- omit in toc -->

[Volker Mische](https://github.com/vmx)

## Table of Contents   <!-- omit in toc -->

- [Install](#install)
  - [In Node.js through npm](#in-nodejs-through-npm)
  - [Browser: Browserify, Webpack, other bundlers](#browser-browserify-webpack-other-bundlers)
  - [In the Browser through `<script>` tag](#in-the-browser-through-script-tag)
- [Usage](#usage)
- [API](#api)
  - [CID.isCID(cid)](#cidiscidcid)
  - [CID.validateCID(cid)](#cidvalidatecidcid)
  - [new CID(version, codec, multihash, [multibaseName])](#new-cidversion-codec-multihash-multibasename)
  - [new CID(baseEncodedString)](#new-cidbaseencodedstring)
  - [new CID(Uint8Array)](#new-ciduint8array)
    - [cid.codec](#cidcodec)
    - [cid.version](#cidversion)
    - [cid.multihash](#cidmultihash)
    - [cid.multibaseName](#cidmultibasename)
    - [cid.bytes](#cidbytes)
    - [cid.prefix](#cidprefix)
    - [cid.toV0()](#cidtov0)
    - [cid.toV1()](#cidtov1)
    - [cid.toBaseEncodedString(base=this.multibaseName)](#cidtobaseencodedstringbasethismultibasename)
    - [cid.toString(base=this.multibaseName)](#cidtostringbasethismultibasename)
    - [cid.equals(cid)](#cidequalscid)
- [Contribute](#contribute)
- [License](#license)

## Install

### In Node.js through npm

```bash
$ npm install --save cids
```

### Browser: Browserify, Webpack, other bundlers

The code published to npm that gets loaded on require is in fact an ES5 transpiled version with the right shims added. This means that you can require it and use with your favourite bundler without having to adjust asset management process.

```js
const CID = require('cids')
```

### In the Browser through `<script>` tag

Loading this module through a script tag will make the ```Cids``` obj available in the global namespace.

```
<script src="https://unpkg.com/cids/dist/index.min.js"></script>
<!-- OR -->
<script src="https://unpkg.com/cids/dist/index.js"></script>
```

## Usage

You can create an instance from a CID string or CID Uint8Array

```js
const CID = require('cids')

const cid = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')

cid.version       // 1
cid.codec         // 'dag-pb'
cid.multibaseName // 'base32'
cid.toString()
// 'bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu'
```

or by specifying the [cid version](https://github.com/multiformats/cid#versions), [multicodec name](https://github.com/multiformats/multicodec/blob/master/table.csv) and [multihash](https://github.com/multiformats/multihash):

```js
const CID = require('cids')
const multihashing = require('multihashing-async')
const bytes = new TextEncoder('utf8').encode('OMG!')

const hash = await multihashing(bytes, 'sha2-256')
const cid = new CID(1, 'dag-pb', hash)
console.log(cid.toString())
// bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu
```

The string form of v1 CIDs defaults to `base32` encoding (v0 CIDs are always `base58btc` encoded). When creating a new instance you can optionally specify the default multibase to use when calling `toBaseEncodedString()` or `toString()`


```js
const cid = new CID(1, 'raw', hash, 'base64')
console.log(cid.toString())
// mAXASIN69ets85WVE0ipva5M5b2mAqAZ8LME08PeAG2MxCSuV
```

If you construct an instance from a valid CID string, the base you provided will be preserved as the default.

```js
// e.g. a base64url encoded CID
const cid = new CID('uAXASIHJSUj5lkfuP5VPWf_VahvhARLRqPkF24QxY-lKaSqvV')
cid.toString()
// uAXASIHJSUj5lkfuP5VPWf_VahvhARLRqPkF24QxY-lKaSqvV
```


## API

### CID.isCID(cid)

Returns true if object is a valid CID instance, false if not valid.

It's important to use this method rather than `instanceof` checks in
order to handle CID objects from different versions of this module.

### CID.validateCID(cid)

Validates the different components (version, codec, multihash, multibaseName) of the CID
instance. Throws an `Error` if not valid.

### new CID(version, codec, multihash, [multibaseName])

`version` must be [either 0 or 1](https://github.com/multiformats/cid#versions).

`codec` must be a string of a valid [registered codec](https://github.com/multiformats/multicodec/blob/master/table.csv).

`multihash` must be a `Uint8Array` instance of a valid [multihash](https://github.com/multiformats/multihash).

`multibaseName` optional string. Must be a valid [multibase](https://github.com/multiformats/multibase/blob/master/multibase.csv) name. Default is `base58btc` for v0 CIDs or `base32` for v1 CIDs.

### new CID(baseEncodedString)

Additionally, you can instantiate an instance from a base encoded
string.

### new CID(Uint8Array)

Additionally, you can instantiate an instance from a `Uint8Array`.

#### cid.codec

Property containing the codec string.

#### cid.version

Property containing the CID version integer.

#### cid.multihash

Property containing the multihash `Uint8Array`.

#### cid.multibaseName

Property containing the default base to use when calling `.toString`

#### cid.bytes

Property containing the full CID encoded as a `Uint8Array`.

#### cid.prefix

Proprety containing a `Uint8Array` of the CID version, codec, and the prefix
section of the multihash.

#### cid.toV0()

Returns the CID encoded in version 0. Only works for `dag-pb` codecs.

Throws if codec is not `dag-pb`.

#### cid.toV1()

Returns the CID encoded in version 1.

#### cid.toBaseEncodedString(base=this.multibaseName)

Returns a base encoded string of the CID. Defaults to the base encoding in `this.multibaseName`.

The value of `this.multibaseName` depends on how the instance was constructed:

1. If the CID was constructed from an object that already had a multibase (a string or an existing CID) then it retains that base.
2. If the CID was constructed from an object that _did not_ have a multibase (a `Uint8Array`, or by passing only version + codec + multihash to the constructor), then `multibaseName` will be `base58btc` for a v0 CID or `base32` for a v1 CID.

#### cid.toString(base=this.multibaseName)

Shorthand for `cid.toBaseEncodedString` described above.

#### cid.equals(cid)

Compare cid instance. Returns true if CID's are identical, false if
otherwise.

## Contribute

[![](https://cdn.rawgit.com/jbenet/contribute-ipfs-gif/master/img/contribute.gif)](https://github.com/ipfs/community/blob/master/contributing.md)

Contributions welcome. Please check out [the issues](https://github.com/multiformats/js-cid/issues).

Check out our [contributing document](https://github.com/ipfs/community/blob/master/CONTRIBUTING_JS.md) for more information on how we work, and about contributing in general. Please be aware that all interactions related to multiformats are subject to the IPFS [Code of Conduct](https://github.com/ipfs/community/blob/master/code-of-conduct.md).

Small note: If editing the Readme, please conform to the [standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## License

MIT


CHANGELOG.md
## [1.1.9](https://github.com/multiformats/js-cid/compare/v1.1.8...v1.1.9) (2021-09-02)



## [1.1.8](https://github.com/multiformats/js-cid/compare/v1.1.7...v1.1.8) (2021-08-24)



## [1.1.7](https://github.com/multiformats/js-cid/compare/v1.1.6...v1.1.7) (2021-06-15)


### Bug Fixes

* add test for v1->v1 conversion ([de28c1f](https://github.com/multiformats/js-cid/commit/de28c1fbbf22ddc47df9193ef5ea73e86fe8d191))



## [1.1.6](https://github.com/multiformats/js-cid/compare/v1.1.5...v1.1.6) (2021-03-02)


### Bug Fixes

* update to aegir 31 and others deps ([#141](https://github.com/multiformats/js-cid/issues/141)) ([bcf2ba2](https://github.com/multiformats/js-cid/commit/bcf2ba2eeeb360da7daa4939459beae76c057993))



## [1.1.5](https://github.com/multiformats/js-cid/compare/v1.1.4...v1.1.5) (2021-01-13)


### Bug Fixes

* fix types and remove typesVersions ([#139](https://github.com/multiformats/js-cid/issues/139)) ([83f3848](https://github.com/multiformats/js-cid/commit/83f38486575c0963f1acf94a3810e0cb918f3981)), closes [/github.com/libp2p/js-libp2p/issues/839#issuecomment-756407072](https://github.com//github.com/libp2p/js-libp2p/issues/839/issues/issuecomment-756407072)
* types to align with multicodec api ([4a5904b](https://github.com/multiformats/js-cid/commit/4a5904bea86a64ab451c9f12c78933e423661b4b))



## [1.1.4](https://github.com/multiformats/js-cid/compare/v1.1.3...v1.1.4) (2020-12-12)


### Bug Fixes

* typing mappings ([f5839b9](https://github.com/multiformats/js-cid/commit/f5839b9788abe231e8d563d64f0fe3db61363c5c)), closes [#131](https://github.com/multiformats/js-cid/issues/131)



## [1.1.3](https://github.com/multiformats/js-cid/compare/v1.1.2...v1.1.3) (2020-12-11)


### Bug Fixes

* use latest multicodec release ([6cc387f](https://github.com/multiformats/js-cid/commit/6cc387ffe12bf29b243eb593545629200a3b5be1))



## [1.1.2](https://github.com/multiformats/js-cid/compare/v1.1.1...v1.1.2) (2020-12-11)


### Bug Fixes

* pin multicodec to version 2.0.4 ([14e55a1](https://github.com/multiformats/js-cid/commit/14e55a155e1135ce9f45ef74a645798480d79df0))
* turn isCID to a type guard ([7413819](https://github.com/multiformats/js-cid/commit/7413819b7b5e20238c87091f22f22665f4cef8eb))



## [1.1.1](https://github.com/multiformats/js-cid/compare/v1.1.0...v1.1.1) (2020-12-11)


### Bug Fixes

* allow CIDs to be compared through deep equality ([127745e](https://github.com/multiformats/js-cid/commit/127745ea90bbeecc80c0bd1f58d24660d832ba14)), closes [#131](https://github.com/multiformats/js-cid/issues/131)



# [1.1.0](https://github.com/multiformats/js-cid/compare/v1.0.2...v1.1.0) (2020-12-11)


### Features

* add ts types with aegir ([0e11f03](https://github.com/multiformats/js-cid/commit/0e11f035c9230e7f6d79c159ace9b80de88cb5eb))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/multiformats/js-cid/compare/v1.0.1...v1.0.2) (2020-10-14)


### Bug Fixes

* improve CID typings ([4948267](https://github.com/multiformats/js-cid/commit/4948267))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/multiformats/js-cid/compare/v1.0.0...v1.0.1) (2020-09-25)


### Bug Fixes

* never return Node.js Buffers ([3c45871](https://github.com/multiformats/js-cid/commit/3c45871)), closes [#126](https://github.com/multiformats/js-cid/issues/126)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/multiformats/js-cid/compare/v0.8.3...v1.0.0) (2020-08-04)


### Bug Fixes

* replace node buffers with uint8arrays ([#117](https://github.com/multiformats/js-cid/issues/117)) ([a7ae250](https://github.com/multiformats/js-cid/commit/a7ae250))


### Features

* expose codec code and allow construction by code ([59469b6](https://github.com/multiformats/js-cid/commit/59469b6)), closes [/github.com/multiformats/js-cid/pull/117#issuecomment-668131658](https://github.com//github.com/multiformats/js-cid/pull/117/issues/issuecomment-668131658)


### BREAKING CHANGES

* - node `Buffer`s have been replaced with `Uint8Array`s
- the `.buffer` property has been renamed to `.bytes` and is now a `Uint8Array`
- the `.prefix` property is now a `Uint8Array`



<a name="0.8.3"></a>
## [0.8.3](https://github.com/multiformats/js-cid/compare/v0.8.2...v0.8.3) (2020-06-19)


### Bug Fixes

* fix support for identity multihash ([#93](https://github.com/multiformats/js-cid/issues/93)) ([51105b6](https://github.com/multiformats/js-cid/commit/51105b6))



<a name="0.8.2"></a>
## [0.8.2](https://github.com/multiformats/js-cid/compare/v0.8.1...v0.8.2) (2020-06-15)


### Features

* support Base36 ([ea5a7f2](https://github.com/multiformats/js-cid/commit/ea5a7f2))



<a name="0.8.1"></a>
## [0.8.1](https://github.com/multiformats/js-cid/compare/v0.8.0...v0.8.1) (2020-05-22)


### Features

* support type predicates ([f7ede61](https://github.com/multiformats/js-cid/commit/f7ede61))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/multiformats/js-cid/compare/v0.7.5...v0.8.0) (2020-03-24)


### Chores

* update deps ([ff2c247](https://github.com/multiformats/js-cid/commit/ff2c247))


### BREAKING CHANGES

* base1 support was removed



<a name="0.7.5"></a>
## [0.7.5](https://github.com/multiformats/js-cid/compare/v0.7.4...v0.7.5) (2020-03-24)


### Features

* add nodejs.util.inspect.custom ([fe953c1](https://github.com/multiformats/js-cid/commit/fe953c1))



<a name="0.7.4"></a>
## [0.7.4](https://github.com/multiformats/js-cid/compare/v0.7.3...v0.7.4) (2020-03-16)


### Bug Fixes

* add buffer ([65681ef](https://github.com/multiformats/js-cid/commit/65681ef))



<a name="0.7.3"></a>
## [0.7.3](https://github.com/multiformats/js-cid/compare/v0.7.2...v0.7.3) (2020-01-24)


### Bug Fixes

* address review requests + ordering ([29e2def](https://github.com/multiformats/js-cid/commit/29e2def))
* update toString to include optional base ([5aff196](https://github.com/multiformats/js-cid/commit/5aff196))


### Features

* more correct type defs + docs ([4eb0c60](https://github.com/multiformats/js-cid/commit/4eb0c60))




test/cid-util.spec.js
/* eslint-env mocha */
/* eslint max-nested-callbacks: ["error", 8] */
'use strict'

const { expect } = require('aegir/utils/chai')
// @ts-ignore
const multihashing = require('multihashing-async')
const CID = require('../src')
const CIDUtil = require('../src/cid-util')
const { fromString: uint8ArrayFromString } = require('uint8arrays/from-string')

describe('CIDUtil', () => {
  /** @type {Uint8Array} */
  let hash

  before(async () => {
    hash = await multihashing(uint8ArrayFromString('abc'), 'sha2-256')
  })

  describe('checkCIDComponents()', () => {
    describe('returns undefined on valid CID', () => {
      it('create from B58Str multihash', () => {
        expect(() => {
          const mhStr = 'QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n'
          const cid = new CID(mhStr)
          const errMsg = CIDUtil.checkCIDComponents(cid)
          expect(errMsg).to.not.exist()
        }).to.not.throw()
      })
      it('create by parts', () => {
        expect(() => {
          const cid = new CID(1, 'dag-cbor', hash)
          const errMsg = CIDUtil.checkCIDComponents(cid)
          expect(errMsg).to.not.exist()
        }).to.not.throw()
      })
    })

    describe('returns non-null error message on invalid inputs', () => {
      const invalid = [
        'hello world',
        'QmaozNR7DZHQK1ZcU9p7QdrshMvXqWK6gpu5rmrkPdT3L',
        uint8ArrayFromString('hello world'),
        uint8ArrayFromString('QmaozNR7DZHQK1ZcU9p7QdrshMvXqWK6gpu5rmrkPdT')
      ]

      invalid.forEach((i) => it(`new CID(${i instanceof Uint8Array ? 'Uint8Array' : 'String'}<${i.toString()}>)`, () => {
        expect(() => {
          const errMsg = CIDUtil.checkCIDComponents(i)
          expect(errMsg).to.exist()
        }).to.not.throw()
      }))

      invalid.forEach((i) => it(`new CID(0, 'dag-pb', ${i instanceof Uint8Array ? 'Uint8Array' : 'String'}<${i.toString()}>)`, () => {
        expect(() => {
          const errMsg = CIDUtil.checkCIDComponents(0)
          expect(errMsg).to.exist()
        }).to.not.throw()
      }))

      invalid.forEach((i) => it(`new CID(1, 'dag-pb', ${i instanceof Uint8Array ? 'Uint8Array' : 'String'}<${i.toString()}>)`, () => {
        expect(() => {
          const errMsg = CIDUtil.checkCIDComponents(1)
          expect(errMsg).to.exist()
        }).to.not.throw()
      }))
    })
  })
})


test/helpers/gen-cid.js
/* eslint-disable no-console */
'use strict'

// CID String: <mbase><version><mcodec><mhash>

const multibase = require('multibase')
const codecs = require('../../src').codecs
// @ts-ignore
const multihashing = require('multihashing-async')
const { fromString: utf8ArrayFromString } = require('uint8arrays/from-string')

async function main () {
  const mh = await multihashing(utf8ArrayFromString('oh, hey!'), 'sha2-256')

  const cid = Uint8Array.of(
    1,
    codecs['dag-pb'],
    ...mh
  )

  const cidStr = multibase.encode('base58btc', cid).toString()

  console.log('CID String (multibase included)')
  console.log(cidStr)
  console.log('CID in hex (multibase not included)')
  console.log(cid.toString())
}

main()
  .catch(err => {
    console.error(err)
    process.exit(0)
  })


src/cid.js
'use strict'

const mh = require('multihashes')
const multibase = require('multibase')

const codecs = require('./codecs')

class CID {
  constructor (codec, version, hash) {
    if (typeof codec === 'string') {
      codec = mh.fromB58String(codec)
    }

    if (Buffer.isBuffer(codec)) {
      this.codec = codecs.protobuf
      this.version = 0
      this.hash = codec
    } else if (typeof codec === 'object') {
      this.codec = codec.codec
      this.version = codec.version
      this.hash = codec.hash
    } else {
      this.codec = codec
      this.version = version
      this.hash = hash
    }
  }

  get buffer () {
    switch (this.version) {
      case 0:
        return this.hash
      case 1:
        return Buffer.concat([
          Buffer(this.version),
          Buffer(this.codec),
          this.hash
        ])
      default:
        throw new Error('unsupported version')
    }
  }

  toString () {
    switch (this.version) {
      case 0:
        return mh.toB58String(this.hash)
      case 1:
        return multibase.encode('base58btc', this.buffer).toString()
      default:
        throw new Error('Unsupported version')
    }
  }

  toJSON () {
    return {
      codec: this.codec,
      version: this.version,
      hash: this.hash
    }
  }

  equals (other) {
    return this.codec === other.codec &&
      this.version === other.version &&
      this.hash.equals(other.hash)
  }
}

CID.codecs = codecs
CID.isCID = (other) => {
  return other.constructor.name === 'CID'
}

module.exports = CID


src/index.d.ts
/**
 * Class representing a CID `<mbase><version><mcodec><mhash>`
 * , as defined in [ipld/cid](https://github.com/multiformats/cid).
 */
declare class CID {
  /**
   * Create a new CID.
   *
   * The algorithm for argument input is roughly:
   * ```
   * if (cid)
   *   -> create a copy
   * else if (str)
   *   if (1st char is on multibase table) -> CID String
   *   else -> bs58 encoded multihash
   * else if (Uint8Array)
   *   if (1st byte is 0 or 1) -> CID
   *   else -> multihash
   * else if (Number)
   *   -> construct CID by parts
   * ```
   *
   * @example
   * new CID(<version>, <codec>, <multihash>, <multibaseName>)
   * new CID(<cidStr>)
   * new CID(<cid.bytes>)
   * new CID(<multihash>)
   * new CID(<bs58 encoded multihash>)
   * new CID(<cid>)
   */
  constructor(
    version: CIDVersion,
    codec: string | number,
    multhash: Uint8Array,
    multibaseName?: string
  );
  constructor(cid: CID|string|Uint8Array);

  /**
   * The version of the CID.
   */
  readonly version: CIDVersion;

  /**
   * The codec of the CID.
   * @deprecated
   */
  readonly codec: string;

  /**
   * The codec of the CID in its number form.
   */
  readonly code: number;

  /**
   * The multihash of the CID.
   */
  readonly multihash: Uint8Array;

  /**
   * Multibase name as string.
   * @deprecated
   */
  readonly multibaseName: string;

  /**
   * The CID as a `Uint8Array`
   */
  readonly bytes: Uint8Array;

  /**
   * The prefix of the CID.
   * @deprecated
   */
  readonly prefix: Uint8Array;

  /**
   * Convert to a CID of version `0`.
   */
  toV0(): CID;

  /**
   * Convert to a CID of version `1`.
   */
  toV1(): CID;

  /**
   * Encode the CID into a string.
   *
   * @param base Base encoding to use.
   */
  toBaseEncodedString(base?: string): string;

  /**
   * Encode the CID into a string.
   */
  toString(base?: string): string;

  /**
   * Serialize to a plain object.
   */
  toJSON(): { codec: string; version: 0 | 1; hash: Uint8Array };

  /**
   * Compare equality with another CID.
   *
   * @param other The other CID.
   */
  equals(other: any): boolean;

  /**
   * Test if the given input is a valid CID object.
   * Throws if it is not.
   *
   * @param other The other CID.
   */
  static validateCID(other: any): void;

  static isCID(mixed: any): mixed is CID;

  static codecs: Record<string, number>;
}

type CIDVersion = 0|1

export = CID


src/cid-util.js
'use strict'

const mh = require('multihashes')

const CIDUtil = {
  /**
   * Test if the given input is a valid CID object.
   * Returns an error message if it is not.
   * Returns undefined if it is a valid CID.
   *
   * @param {any} other
   * @returns {string}
   */
  checkCIDComponents: function (other) {
    if (other == null) {
      return 'null values are not valid CIDs'
    }

    if (!(other.version === 0 || other.version === 1)) {
      return 'Invalid version, must be a number equal to 1 or 0'
    }

    if (typeof other.codec !== 'string') {
      return 'codec must be string'
    }

    if (other.version === 0) {
      if (other.codec !== 'dag-pb') {
        return "codec must be 'dag-pb' for CIDv0"
      }
      if (other.multibaseName !== 'base58btc') {
        return "multibaseName must be 'base58btc' for CIDv0"
      }
    }

    if (!(other.multihash instanceof Uint8Array)) {
      return 'multihash must be a Uint8Array'
    }

    try {
      mh.validate(other.multihash)
    } catch (err) {
      let errorMsg = err.message
      if (!errorMsg) { // Just in case mh.validate() throws an error with empty error message
        errorMsg = 'Multihash validation failed'
      }
      return errorMsg
    }
  }
}

module.exports = CIDUtil


src/index.js
'use strict'

const mh = require('multihashes')
const multibase = require('multibase')
const multicodec = require('multicodec')
const CIDUtil = require('./cid-util')
const uint8ArrayConcat = require('uint8arrays/concat')
const uint8ArrayToString = require('uint8arrays/to-string')
const uint8ArrayEquals = require('uint8arrays/equals')

const codecs = multicodec.nameToCode
const codecInts = /** @type {CodecName[]} */(Object.keys(codecs)).reduce((p, name) => {
  p[codecs[name]] = name
  return p
}, /** @type {Record<CodecCode, CodecName>} */({}))

const symbol = Symbol.for('@ipld/js-cid/CID')

/**
 * @typedef {Object} SerializedCID
 * @property {string} codec
 * @property {number} version
 * @property {Uint8Array} hash
 */
/**
 * @typedef {0|1} CIDVersion
 * @typedef {import('multibase').BaseNameOrCode} BaseNameOrCode
 * @typedef {import('multicodec').CodecName} CodecName
 * @typedef {import('multicodec').CodecCode} CodecCode
 */

/**
 * Class representing a CID `<mbase><version><mcodec><mhash>`
 * , as defined in [ipld/cid](https://github.com/multiformats/cid).
 *
 * @class CID
 */
class CID {
  /**
   * Create a new CID.
   *
   * The algorithm for argument input is roughly:
   * ```
   * if (cid)
   *   -> create a copy
   * else if (str)
   *   if (1st char is on multibase table) -> CID String
   *   else -> bs58 encoded multihash
   * else if (Uint8Array)
   *   if (1st byte is 0 or 1) -> CID
   *   else -> multihash
   * else if (Number)
   *   -> construct CID by parts
   * ```
   *
   * @param {CIDVersion | string | Uint8Array | CID} version
   * @param {string|number} [codec]
   * @param {Uint8Array} [multihash]
   * @param {string} [multibaseName]
   *
   * @example
   * new CID(<version>, <codec>, <multihash>, <multibaseName>)
   * new CID(<cidStr>)
   * new CID(<cid.bytes>)
   * new CID(<multihash>)
   * new CID(<bs58 encoded multihash>)
   * new CID(<cid>)
   */
  constructor (version, codec, multihash, multibaseName) {
    // We have below three blank field accessors only because
    // otherwise TS will not pick them up if done after assignemnts

    /**
     * The version of the CID.
     *
     * @type {CIDVersion}
     */
    // eslint-disable-next-line no-unused-expressions
    this.version

    /**
     * The codec of the CID.
     *
     * @deprecated
     * @type {CodecName}
     */
    // eslint-disable-next-line no-unused-expressions
    this.codec

    /**
     * The multihash of the CID.
     *
     * @type {Uint8Array}
     */
    // eslint-disable-next-line no-unused-expressions
    this.multihash

    Object.defineProperty(this, symbol, { value: true })
    if (CID.isCID(version)) {
      // version is an exising CID instance
      const cid = /** @type {CID} */(version)
      this.version = cid.version
      this.codec = cid.codec
      this.multihash = cid.multihash
      // Default guard for when a CID < 0.7 is passed with no multibaseName
      // @ts-ignore
      this.multibaseName = cid.multibaseName || (cid.version === 0 ? 'base58btc' : 'base32')
      return
    }

    if (typeof version === 'string') {
      // e.g. 'base32' or false
      const baseName = multibase.isEncoded(version)
      if (baseName) {
        // version is a CID String encoded with multibase, so v1
        const cid = multibase.decode(version)
        this.version = /** @type {CIDVersion} */(parseInt(cid[0].toString(), 16))
        this.codec = multicodec.getCodec(cid.slice(1))
        this.multihash = multicodec.rmPrefix(cid.slice(1))
        this.multibaseName = baseName
      } else {
        // version is a base58btc string multihash, so v0
        this.version = 0
        this.codec = 'dag-pb'
        this.multihash = mh.fromB58String(version)
        this.multibaseName = 'base58btc'
      }
      CID.validateCID(this)
      Object.defineProperty(this, 'string', { value: version })
      return
    }

    if (version instanceof Uint8Array) {
      const v = parseInt(version[0].toString(), 16)
      if (v === 1) {
        // version is a CID Uint8Array
        const cid = version
        this.version = v
        this.codec = multicodec.getCodec(cid.slice(1))
        this.multihash = multicodec.rmPrefix(cid.slice(1))
        this.multibaseName = 'base32'
      } else {
        // version is a raw multihash Uint8Array, so v0
        this.version = 0
        this.codec = 'dag-pb'
        this.multihash = version
        this.multibaseName = 'base58btc'
      }
      CID.validateCID(this)
      return
    }

    // otherwise, assemble the CID from the parameters

    this.version = version

    if (typeof codec === 'number') {
      // @ts-ignore
      codec = codecInts[codec]
    }

    this.codec = /** @type {CodecName} */ (codec)

    this.multihash = /** @type {Uint8Array} */ (multihash)

    /**
     * Multibase name as string.
     *
     * @deprecated
     * @type {string}
     */
    this.multibaseName = multibaseName || (version === 0 ? 'base58btc' : 'base32')

    CID.validateCID(this)
  }

  /**
   * The CID as a `Uint8Array`
   *
   * @returns {Uint8Array}
   *
   */
  get bytes () {
    // @ts-ignore
    let bytes = this._bytes

    if (!bytes) {
      if (this.version === 0) {
        bytes = this.multihash
      } else if (this.version === 1) {
        const codec = multicodec.getCodeVarint(this.codec)
        bytes = uint8ArrayConcat([
          [1], codec, this.multihash
        ], 1 + codec.byteLength + this.multihash.byteLength)
      } else {
        throw new Error('unsupported version')
      }

      // Cache this Uint8Array so it doesn't have to be recreated
      Object.defineProperty(this, '_bytes', { value: bytes })
    }

    return bytes
  }

  /**
   * The prefix of the CID.
   *
   * @returns {Uint8Array}
   */
  get prefix () {
    const codec = multicodec.getCodeVarint(this.codec)
    const multihash = mh.prefix(this.multihash)
    const prefix = uint8ArrayConcat([
      [this.version], codec, multihash
    ], 1 + codec.byteLength + multihash.byteLength)

    return prefix
  }

  /**
   * The codec of the CID in its number form.
   *
   * @returns {CodecCode}
   */
  get code () {
    return codecs[this.codec]
  }

  /**
   * Convert to a CID of version `0`.
   *
   * @returns {CID}
   */
  toV0 () {
    if (this.codec !== 'dag-pb') {
      throw new Error('Cannot convert a non dag-pb CID to CIDv0')
    }

    const { name, length } = mh.decode(this.multihash)

    if (name !== 'sha2-256') {
      throw new Error('Cannot convert non sha2-256 multihash CID to CIDv0')
    }

    if (length !== 32) {
      throw new Error('Cannot convert non 32 byte multihash CID to CIDv0')
    }

    return new CID(0, this.codec, this.multihash)
  }

  /**
   * Convert to a CID of version `1`.
   *
   * @returns {CID}
   */
  toV1 () {
    if (this.version === 0) {
      return new CID(1, this.codec, this.multihash, 'base32')
    }

    return new CID(1, this.codec, this.multihash, this.multibaseName)
  }

  /**
   * Encode the CID into a string.
   *
   * @param {BaseNameOrCode} [base=this.multibaseName] - Base encoding to use.
   * @returns {string}
   */
  toBaseEncodedString (base = this.multibaseName) {
    // @ts-ignore non enumerable cache property
    if (this.string && this.string.length !== 0 && base === this.multibaseName) {
      // @ts-ignore non enumerable cache property
      return this.string
    }
    let str
    if (this.version === 0) {
      if (base !== 'base58btc') {
        throw new Error('not supported with CIDv0, to support different bases, please migrate the instance do CIDv1, you can do that through cid.toV1()')
      }
      str = mh.toB58String(this.multihash)
    } else if (this.version === 1) {
      str = uint8ArrayToString(multibase.encode(base, this.bytes))
    } else {
      throw new Error('unsupported version')
    }
    if (base === this.multibaseName) {
      // cache the string value
      Object.defineProperty(this, 'string', { value: str })
    }
    return str
  }

  /**
   * CID(QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n)
   *
   * @returns {string}
   */
  [Symbol.for('nodejs.util.inspect.custom')] () {
    return 'CID(' + this.toString() + ')'
  }

  /**
   * Encode the CID into a string.
   *
   * @param {BaseNameOrCode} [base=this.multibaseName] - Base encoding to use.
   * @returns {string}
   */
  toString (base) {
    return this.toBaseEncodedString(base)
  }

  /**
   * Serialize to a plain object.
   *
   * @returns {SerializedCID}
   */
  toJSON () {
    return {
      codec: this.codec,
      version: this.version,
      hash: this.multihash
    }
  }

  /**
   * Compare equality with another CID.
   *
   * @param {CID} other
   * @returns {boolean}
   */
  equals (other) {
    return this.codec === other.codec &&
      this.version === other.version &&
      uint8ArrayEquals(this.multihash, other.multihash)
  }

  /**
   * Test if the given input is a valid CID object.
   * Throws if it is not.
   *
   * @param {any} other - The other CID.
   * @returns {void}
   */
  static validateCID (other) {
    const errorMsg = CIDUtil.checkCIDComponents(other)
    if (errorMsg) {
      throw new Error(errorMsg)
    }
  }

  /**
   * Check if object is a CID instance
   *
   * @param {any} value
   * @returns {value is CID}
   */
  static isCID (value) {
    return value instanceof CID || Boolean(value && value[symbol])
  }
}

CID.codecs = codecs

module.exports = CID
